# Exploring Ebay Car Sales Data from eBay Kleinanzeigen

## Introduction:

This analysis will focus on a classifieds section of the German eBay website, eBay Kleinanzeigen.

The dataset was scraped from the web and can be found at [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).

for this analysis report, the dataset we're using here was modified as per the following:
1. A 50,000 data points were sampled from the full data set. *(This was intentional for the purpose of learning at the training website [dataquest.io](https://dataquest.io))*
2. The original dataset is a clean one, so for the purpose of learning, the dataset we're working on has been dirtied a bit to more closely resemble what we can expect from a scraped dataset in real life.

## Goals:

The aim of this analysis is to:
1. Clean the data.
2. Analyze the included used car listings.
3. Be familiar with some of the unique benefits jupyter notebook provides for pandas.

## Columns:

Below, is a list of each column in the dataset, along with a brief description.

1. `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
2. `name` - Name of the car.
3. `seller` - Whether the seller is private or a dealer.
4. `offerType` - The type of listing
5. `price` - The price on the ad to sell the car.
6. `abtest` - Whether the listing is included in an A/B test.
7. `vehicleType` - The vehicle Type.
8. `yearOfRegistration` - The year in which the car was first registered.
9. `gearbox` - The transmission type.
10. `powerPS` - The power of the car in PS.
11. `model` - The car model name.
12. `kilometer` - How many kilometers the car has driven.
13. `monthOfRegistration` - The month in which the car was first registered.
14. `fuelType` - What type of fuel the car uses.
15. `brand` - The brand of the car.
16. `notRepairedDamage` - If the car has a damage which is not yet repaired.
17. `dateCreated` - The date on which the eBay listing was created.
18. `nrOfPictures` - The number of pictures in the ad.
19. `postalCode` - The postal code for the location of the vehicle.
20. `lastSeenOnline` - When the crawler saw this ad last online.

## Preparation code:

In [1]:
#importing libraries and tools
import numpy as np
import pandas as pd

#opening the dataset into pandas using the Latin-1 encoding
#since the default encoding seems to give an error
autos = pd.read_csv('autos.csv', encoding='Latin-1')


In [2]:
#A neat fature of jupyter notebook is its ability to render
#the first few and the last few values of any pandas object

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

## Data Exploration and Pre-Analysis Cleaning:

In here, we will explore the basic information about this dataset, as well as clean/fix some issues with it.

In [3]:
autos.info()
print('\n')
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

From the above, we notice that the columns uses camelCase, as well as only few of the columns are read as int64, where the rest (dates, prices, some other numbers, etc.) are objects. we can explore more like below:

In [4]:
columnsToCheck = ['seller',
                  'offerType',
                  'abtest',
                  'vehicleType',
                  'gearbox',
                  'monthOfRegistration',
                  'fuelType',
                  'brand',
                  'notRepairedDamage'
                 ]

for col in columnsToCheck:
    print(col, ':\n', autos[col].unique(), '\n')

seller :
 ['privat' 'gewerblich'] 

offerType :
 ['Angebot' 'Gesuch'] 

abtest :
 ['control' 'test'] 

vehicleType :
 ['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere'] 

gearbox :
 ['manuell' 'automatik' nan] 

monthOfRegistration :
 [ 3  6  7  4  8 12 10  0  9 11  5  2  1] 

fuelType :
 ['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere'] 

brand :
 ['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia'] 

notRepairedDamage :
 ['nein' nan 'ja'] 



We can see from the above that we do have some missing data in the fields `vehicleType`, `fuelType`, and `notRepairedDamage`.

In [5]:
#Analysing NaN values percentages
d = {}
for col in columnsToCheck:
    temp = autos[col].isnull()
    i = len(autos.loc[temp,col])
    d[col] = i / 50000 * 100

print(d)

{'seller': 0.0, 'gearbox': 5.36, 'offerType': 0.0, 'vehicleType': 10.190000000000001, 'abtest': 0.0, 'monthOfRegistration': 0.0, 'notRepairedDamage': 19.658, 'fuelType': 8.964, 'brand': 0.0}


From the above values, we notice that the NaN values do not go over 20%.

###  Columns names

We will fix the current format for the columns to be more descriptive:
1. The column names will be changed from a cameCase, to the Python's preferred snakecase.
2. we will reword some of the column names based on the data dictionary.

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

As per the above results, we have fixed the column names to a neat snakecase with Pandas.

### Removal of non-useful information

In some datasets, we can find some columns that contains non-useful data (e.g. text columns where all or almost all values are the same, which are not useful for analysis). In here, we will do a bit more of exploration to figure this out.

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

#### A.

From the above stats, we can see that the column `nr_of_pictures` have the same value, which is 0.0. We can confirm this from the below code:

In [9]:
autos['nr_of_pictures'].unique()

array([0])

This denotes that the `nr_of_pictures` is not very useful for our analysis.

#### B.

Few of the columns may have 2 values only (like true or flase values), such as `unrepaired_damage`, and `seller` columns.

In [10]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [11]:
autos['seller'].unique()

array(['privat', 'gewerblich'], dtype=object)

Both columns can have just 2 values, and for `unrepaired_damage` column, we have nan values as well.

#### C.

Columns `seller` and `offer_type` have almost all of the values the same.

In [12]:
autos['seller'].describe()

count      50000
unique         2
top       privat
freq       49999
Name: seller, dtype: object

In [13]:
autos['offer_type'].describe()

count       50000
unique          2
top       Angebot
freq        49999
Name: offer_type, dtype: object

#### D.

Some columns are still not clear in its data, since they're stored as text. We will convert them as they should be numeric. These columns are are `price` and `odometer` columns.

### Converting text to numeric data

We will work on `odometer` and `price` columns first, since they're stored as text and not in a numeric type.

In [14]:
autos['odometer'].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [15]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

The `odometer` column contains 'km' string at the end of each entry, while the `price` column contains '$' at the beginning of each entry.

In [16]:
#odometer series conversion
autos['odometer'] = autos['odometer'].str[:-2]
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].astype(int)

#price series convertion
autos['price'] = autos['price'].str[1:]
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)

#converting column names
autos.rename(columns={'odometer':'odometer_km', 'price':'price_usd'}, inplace=True)

In [17]:
#checking the heads again
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [18]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price_usd             50000 non-null int64
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(7), 

Now we can see both columns are fixed (datatype wise). New column names are `price_usd` and `odometer_km`.

### Finding outliers

In this step, we will look for data that doesn't look right. We will analyze the `odometer_km` and `price_usd` columns for any abnormal data entries (i.e. unrealistic high/low values).

In [19]:
#column `odometer_km` analysis
print('Min: ', autos['odometer_km'].min())
print('Max: ', autos['odometer_km'].max())

print('Unique values count: ', autos['odometer_km'].unique().shape, end='\n\n')

print('Detailed description: ')
print(autos['odometer_km'].describe(),end='\n\n')
print('Value counts: ')
print(autos['odometer_km'].value_counts().sort_index())

Min:  5000
Max:  150000
Unique values count:  (13,)

Detailed description: 
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Value counts: 
5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64


The `odometer_km` column's data shows that we only have 13 unique values. This can be helpful for our analysis as the ads are already categorized with few values for `odmeter_km` column.

This can denote that, when ads are placed, the `odometer_km` box is a combo box, with only 13 values.

In [20]:
#column `price_usd` analysis
print('Min: ', autos['price_usd'].min())
print('Max: ', autos['price_usd'].max())

print('Unique values count: ', autos['price_usd'].unique().shape, end='\n\n')

print('Detailed description: ')
print(autos['price_usd'].describe(),end='\n\n')
print('Value counts (Lowest 5): ')
print(autos['price_usd'].value_counts().sort_index().head(), end='\n\n')

print('Value counts (Highest 5): ')
print(autos['price_usd'].value_counts().sort_index(ascending=False).head())

Min:  0
Max:  99999999
Unique values count:  (2357,)

Detailed description: 
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

Value counts (Lowest 5): 
0    1421
1     156
2       3
3       1
5       2
Name: price_usd, dtype: int64

Value counts (Highest 5): 
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price_usd, dtype: int64


We can see from the above the the `price_usd` column contains some outliers, and they should be removed. We first try and see cars above $ 1,000,000.

In [21]:
autos.loc[autos['price_usd']>100000, :]

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
1878   2016-03-12 16:58:37                                  Porsche_911_Turbo   
2454   2016-03-21 22:51:29                                    Porsche_911_GT3   
2751   2016-03-15 10:52:35                               Porsche_911___993_4S   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7402   2016-03-22 19:48:09  Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
8232   2016-04-01 21:50:47               Porsche_993_S_Schalter_BRD_neuwertig   
10500  2016-03-17 12:56:38                                        Porsche_991   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
11433  2016-03-30 21:44:21  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
14268  2016-04-01 12:52:27   Porsche_997_GT3_Approved_Lift_Unfallfrei_Mj.2011   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
16964  2016-04-01 16:46:18                    Bentley_Continental_Supersports   
17140  2016-03-05 17:39:22  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
17540  2016-03-30 00:49:49   Porsche_930_Turbo_classic_analytics_Gutachten_2+   
18509  2016-03-27 18:43:02                                    Porsche_991_GT3   
20351  2016-03-22 17:47:59                                    BMW_Z8_roadster   
21783  2016-03-26 22:06:24               Porsche_911_Carrera_4S_Cabriolet_PDK   
22060  2016-03-09 00:58:24  Tesla_Model_X90D_Autopilot_Leder_AHK_Kaltwette...   
22673  2016-03-05 18:40:13  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
28090  2016-03-08 19:50:06  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
29286  2016-03-22 15:58:56  Porsche_911_Turbo_PDK__Sport_Chrono_Garantie_D...   
32185  2016-03-25 12:42:09  Porsche_997_GT3_MKII_EIN_SAMMLERSTÜCK_IN_VOLLA...   
32840  2016-03-22 08:55:27  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
33638  2016-03-24 12:54:19                                 Porsche_911_GT3_RS   
33884  2016-03-26 21:55:12               Porsche_911_Carrera_4S_Cabriolet_PDK   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
36818  2016-03-27 18:37:37                                        Porsche_991   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
37840  2016-03-21 10:50:12                                        Porsche_997   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
38814  2016-03-22 16:53:44  Porsche_Porsche_964_TURBO_S_X33_WLS_ab_Werk_35...   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
40918  2016-03-20 18:40:05                             Porsche_911_991_GT3_RS   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
43282  2016-04-03 18:42:28               Porsche_911_Carrera_4S_Cabriolet_PDK   
43668  2016-03-16 18:47:26  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
44406  2016-03-15 14:47:26  Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...   
45387  2016-03-18 16:41:22  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6

In [22]:
print(len(autos.loc[autos['price_usd']>1000000, :]))

11


Since they are only 11, we can go ahead and remove them:

In [23]:
autos.drop(autos[autos['price_usd'] > 1000000].index, inplace=True)

We can also see those values with unrealistic values, such as zero dollars, 999999 dollars, and 1 dollar, etc:

In [24]:
print(len(autos.loc[autos['price_usd']==0, :]))
print(len(autos.loc[autos['price_usd']==999999, :]))
print(len(autos.loc[autos['price_usd']==1, :]))

1421
2
156


We can remove these from our dataset as well, since they won't help in the analysis:

In [25]:
autos.drop(autos[autos['price_usd'] == 0].index, inplace=True)
autos.drop(autos[autos['price_usd'] == 999999].index, inplace=True)
autos.drop(autos[autos['price_usd'] == 1].index, inplace=True)

We can analyze the `price_usd` columns again now:

In [26]:
#column `price_usd` analysis
print('Min: ', autos['price_usd'].min())
print('Max: ', autos['price_usd'].max())

print('Unique values count: ', autos['price_usd'].unique().shape, end='\n\n')

print('Detailed description: ')
print(autos['price_usd'].describe(),end='\n\n')
print('Value counts (Lowest 5): ')
print(autos['price_usd'].value_counts().sort_index().head(), end='\n\n')

print('Value counts (Highest 5): ')
print(autos['price_usd'].value_counts().sort_index(ascending=False).head())

Min:  2
Max:  999990
Unique values count:  (2346,)

Detailed description: 
count     48410.000000
mean       5928.444350
std       10131.378228
min           2.000000
25%        1250.000000
50%        3000.000000
75%        7496.000000
max      999990.000000
Name: price_usd, dtype: float64

Value counts (Lowest 5): 
2    3
3    1
5    2
8    1
9    1
Name: price_usd, dtype: int64

Value counts (Highest 5): 
999990    1
350000    1
345000    1
299000    1
295000    1
Name: price_usd, dtype: int64


From the above, we can have an overview of the remaining values. By looking at the highest 5 values, we can determine tat about 350000 dollars is the highest value (we can ignore the 999990 dollars as it looks unrealistic.

As well, by looking at the lowest 5 values, we seem to still find one-digit values. We can try and analyze the lowest values more:

In [27]:
autos['price_usd'].sort_values().head(200)

1171       2
24413      2
30678      2
33429      3
8560       5
5375       5
28372      8
31054      9
27658     10
29246     10
35009     10
44597     10
42186     10
33893     10
17079     10
9255      11
35555     11
3416      12
18443     12
15747     12
4598      13
5132      13
19632     14
45157     15
40856     15
27813     17
38552     17
30587     17
8570      18
37141     20
        ... 
26171     99
46896     99
19082     99
10507     99
8583      99
11640     99
38067     99
3221      99
38682     99
1455      99
46818     99
22385     99
42802     99
33405     99
40737     99
39455    100
28302    100
27789    100
30734    100
13741    100
34359    100
34951    100
8306     100
27761    100
3878     100
34344    100
16601    100
19014    100
26236    100
13590    100
Name: price_usd, Length: 200, dtype: int64

For the purpose of our analyss, we can analyze if we can go ahead and drop the values of 100 and less:

In [28]:
len(autos.loc[autos['price_usd'] <= 100,:])

319

A records count of 319 should be fine if they get removed, since they won't affect the analysis much:

In [29]:
autos.drop(autos[autos['price_usd'] <=100].index, inplace=True)

Now, we perform our analysis one more time:

In [30]:
#column `price_usd` analysis
print('Min: ', autos['price_usd'].min())
print('Max: ', autos['price_usd'].max())

print('Unique values count: ', autos['price_usd'].unique().shape, end='\n\n')

print('Detailed description: ')
print(autos['price_usd'].describe(),end='\n\n')
print('Value counts (Lowest 5): ')
print(autos['price_usd'].value_counts().sort_index().head(), end='\n\n')

print('Value counts (Highest 5): ')
print(autos['price_usd'].value_counts().sort_index(ascending=False).head())

Min:  110
Max:  999990
Unique values count:  (2310,)

Detailed description: 
count     48091.000000
mean       5967.287476
std       10153.656015
min         110.000000
25%        1250.000000
50%        3099.000000
75%        7500.000000
max      999990.000000
Name: price_usd, dtype: float64

Value counts (Lowest 5): 
110     3
111     2
115     2
117     1
120    39
Name: price_usd, dtype: int64

Value counts (Highest 5): 
999990    1
350000    1
345000    1
299000    1
295000    1
Name: price_usd, dtype: int64


By looking at the numbers, we can see that most of the values are below 7500 dollars.

In [31]:
print(autos['price_usd'].unique().shape)
print(autos['odometer_km'].unique().shape)

(2310,)
(13,)


As well as the total unique values are few compared to the total number of records.

Now in total, we have removed less than 4% of the records.

### Dates columns

In the dataset, the columns which contains dates are as below:

1. `date_crawled` at index 0 - When this ad was first crawled. All field-values are taken from this date (Added by the crawler).
2. `registration_year` at index 7 - The year in which the car was first registered (from the website).
3. `registration_month` at index 12 - The month in which the car was first registered (from the website).
4. `ad_created` at index 16 - The date on which the eBay listing was created (from the website).
5. `last_seen` at index 19 - When the crawler saw this ad last online (Added by the crawler).

The first step involves identifyting the current format of each column, and convert any string values, to numeric values so we can understand it quantitatively.

In [32]:
print(autos['date_crawled'].dtypes)
print(autos['registration_month'].dtypes)
print(autos['registration_year'].dtypes)
print(autos['ad_created'].dtypes)
print(autos['last_seen'].dtypes)

object
int64
int64
object
object


The columns `registration_month` and `registration_year` are already read by Pandas as a numeric data type. We're going to work on the rest of the colunmns to convert them into a numeric type.

In [33]:
#exploring columns
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We notice from the above that the first 10 characters of each value represents the day. Therefore, we can analyze the columns like below to see the distribution of the dates along the records.

#### `date_crawled`

In [34]:
print(autos['date_crawled'].str[0:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025389
2016-03-06    0.014057
2016-03-07    0.035953
2016-03-08    0.033229
2016-03-09    0.033000
2016-03-10    0.032355
2016-03-11    0.032646
2016-03-12    0.036951
2016-03-13    0.015720
2016-03-14    0.036639
2016-03-15    0.034227
2016-03-16    0.029444
2016-03-17    0.031503
2016-03-18    0.012871
2016-03-19    0.034726
2016-03-20    0.037762
2016-03-21    0.037283
2016-03-22    0.032917
2016-03-23    0.032293
2016-03-24    0.029444
2016-03-25    0.031399
2016-03-26    0.032231
2016-03-27    0.031128
2016-03-28    0.034996
2016-03-29    0.034102
2016-03-30    0.033790
2016-03-31    0.031835
2016-04-01    0.033686
2016-04-02    0.035599
2016-04-03    0.038614
2016-04-04    0.036556
2016-04-05    0.013100
2016-04-06    0.003161
2016-04-07    0.001393
Name: date_crawled, dtype: float64


From the above, we can observe that:
1. The craweler took about a month to collect the data.
2. The craweler takes a bulk records each day, while on some days, it takes few data comparing the other days.
3. The crawling activity began to stop on 6th of April by taking only few records.

#### `ad_created`

In [50]:
autos['ad_created'].str[0:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000042
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000042
2016-02-05    0.000042
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000042
2016-02-14    0.000042
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000042
2016-02-19    0.000062
2016-02-20    0.000042
2016-02-21    0.000062
                ...   
2016-03-09    0.033083
2016-03-10    0.032064
2016-03-11    0.032958
2016-03-12    0.036784
2016-03-13    0.017093
2016-03-14    0.035266
2016-03-15    0.033956
2016-03-16    0.029943
2016-03-17    0.031170
2016-03-18    0.013558
2016-03-19    0.033582
2016-03-20    0.037845
2016-03-21 

The above shows just 76 records. Lets get the last 50 to see the distribution.

In [52]:
autos['ad_created'].str[0:10].value_counts(normalize=True, dropna=False).sort_index().tail(50)

2016-02-18    0.000042
2016-02-19    0.000062
2016-02-20    0.000042
2016-02-21    0.000062
2016-02-22    0.000021
2016-02-23    0.000083
2016-02-24    0.000042
2016-02-25    0.000062
2016-02-26    0.000042
2016-02-27    0.000125
2016-02-28    0.000208
2016-02-29    0.000166
2016-03-01    0.000104
2016-03-02    0.000104
2016-03-03    0.000873
2016-03-04    0.001497
2016-03-05    0.022936
2016-03-06    0.015325
2016-03-07    0.034684
2016-03-08    0.033229
2016-03-09    0.033083
2016-03-10    0.032064
2016-03-11    0.032958
2016-03-12    0.036784
2016-03-13    0.017093
2016-03-14    0.035266
2016-03-15    0.033956
2016-03-16    0.029943
2016-03-17    0.031170
2016-03-18    0.013558
2016-03-19    0.033582
2016-03-20    0.037845
2016-03-21    0.037512
2016-03-22    0.032730
2016-03-23    0.032127
2016-03-24    0.029382
2016-03-25    0.031524
2016-03-26    0.032293
2016-03-27    0.031045
2016-03-28    0.035100
2016-03-29    0.034060
2016-03-30    0.033603
2016-03-31    0.031877
2016-04-01 

We observe from the above that:
1. Ads created mostly after the 4th of March.
2. The distibution of the records kinda syncs with `date_crawled` records, as few days where `date_crawled` had less distribution, `ad_created` had less distribution as well. This may denote that the site was down at that time, and the crawler was just fine.

#### `last_seen`

In [36]:
print(autos['last_seen'].str[0:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001081
2016-03-06    0.004304
2016-03-07    0.005386
2016-03-08    0.007299
2016-03-09    0.009565
2016-03-10    0.010646
2016-03-11    0.012414
2016-03-12    0.023830
2016-03-13    0.008879
2016-03-14    0.012622
2016-03-15    0.015824
2016-03-16    0.016427
2016-03-17    0.028072
2016-03-18    0.007319
2016-03-19    0.015762
2016-03-20    0.020669
2016-03-21    0.020565
2016-03-22    0.021376
2016-03-23    0.018590
2016-03-24    0.019713
2016-03-25    0.019110
2016-03-26    0.016698
2016-03-27    0.015512
2016-03-28    0.020794
2016-03-29    0.022333
2016-03-30    0.024641
2016-03-31    0.023851
2016-04-01    0.022873
2016-04-02    0.024849
2016-04-03    0.025140
2016-04-04    0.024516
2016-04-05    0.125159
2016-04-06    0.221954
2016-04-07    0.132228
Name: last_seen, dtype: float64


The increasing number of distribution in the last 3 days (5th till 7th of April) could mean that the crawler was verifying the current record held, and this may explain the slowness in records creation during these days.

We can observe from the analysis of the 3 columns that no column have any missing values

#### `registration_year`

In [62]:
print(autos['registration_year'].value_counts(dropna=False, normalize=True).sort_index())

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000042
1910    0.000042
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000083
1957    0.000042
1958    0.000083
1959    0.000125
1960    0.000437
1961    0.000125
1962    0.000083
1963    0.000166
1964    0.000250
          ...   
2000    0.064024
2001    0.054626
2002    0.051465
2003    0.055998
2004    0.056123
2005    0.060489
2006    0.055478
2007    0.047244
2008    0.045934
2009    0.043251
2010    0.033000
2011    0.033624
2012    0.027198
2013    0.016635
2014    0.013766
2015    0.007902
2016    0.024870
2017    0.028737
2018    0.009732
2019    0.000021
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000062
5911    0.000021
6200    0.000021
8888    0.0000

We can observe unusual values here, so we can go ahead and remove them.

In [63]:
autos.drop(autos[autos['registration_year'] > 2019].index, inplace=True)
autos.drop(autos[autos['registration_year'] < 1910].index, inplace=True)
print(autos.shape)

(48072, 20)


In [64]:
print(autos['registration_year'].value_counts(dropna=False, normalize=True).sort_index())

1910    0.000042
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000083
1957    0.000042
1958    0.000083
1959    0.000125
1960    0.000437
1961    0.000125
1962    0.000083
1963    0.000166
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000541
1968    0.000541
          ...   
1990    0.006823
1991    0.006969
1992    0.007593
1993    0.008654
1994    0.012814
1995    0.024484
1996    0.028020
1997    0.039690
1998    0.048323
1999    0.059744
2000    0.064050
2001    0.054647
2002    0.051485
2003    0.056020
2004    0.056145
2005    0.060513
2006    0.055500
2007    0.047262
2008    0.045952
2009    0.043268
2010    0.033013
2011    0.033637
2012    0.027209
2013    0.016642
2014    0.013771
2015    0.007905
2016    0.024879
2017    0.0287

In [65]:
autos['registration_year'].describe()

count    48072.000000
mean      2003.505180
std          7.507013
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

From the above, we can estimate that most of the cars are registered between 1996 and 2010 (using the mean and the standard deviation).

In [73]:
autos['registration_year'].value_counts().sort_values(ascending=False).head()

2000    3079
2005    2909
1999    2872
2004    2699
2003    2693
Name: registration_year, dtype: int64

From the above, we can confirm our statement, and also note that most of the cars were registered in 2000.

### Identifying brands

Let us look at the `brand` column within the dataset, which holds the car brand for each ad.

In [79]:
print(autos['brand'].unique())
print('length: ', len(autos['brand'].unique()))

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']
length:  40


As expected, brands are limited in count comapred to the number of ads in the dataset. These brands contian no numeric values, and shall be treated as strings if required.

In [85]:
#Identify top 20 brands
autos['brand'].value_counts().sort_values(ascending=False).head(20)

volkswagen        10238
bmw                5241
opel               5196
mercedes_benz      4625
audi               4143
ford               3334
renault            2294
peugeot            1416
fiat               1244
seat                910
skoda               774
nissan              737
mazda               732
smart               690
citroen             679
toyota              611
hyundai             479
sonstige_autos      450
volvo               434
mini                416
Name: brand, dtype: int64

From the above, we can see that most ads are on the top 9 brands displayed above (over 1000 ads per brand). We can start analyzing those.

In [100]:
brand_price_mean_dct = {}

for item in autos['brand'].value_counts().sort_values(ascending=False).head(9).index:
    brand_price_mean_dct[item] = autos.loc[autos['brand'] == item, 'price_usd'].mean()

print(brand_price_mean_dct)

{'volkswagen': 5478.835124047666, 'peugeot': 3095.3679378531074, 'mercedes_benz': 8572.600216216217, 'ford': 3781.123275344931, 'renault': 2463.1992153443766, 'audi': 9268.353608496258, 'bmw': 8313.271513070025, 'fiat': 2833.099678456592, 'opel': 2984.8720169361045}


From the above, the brands audi, mercedex_benz, and bmw, have the highest mean values for the price.

On the other hand, the brands renault, fiat, and opel have the lowest mean values for the price.

#### Mileage and price link

Let us calculate the average mileage for the brands we got to see if there is any link between it and the mean price.

In [102]:
brand_mileage_mean_dct = {}

for item in autos['brand'].value_counts().sort_values(ascending=False).head(9).index:
    brand_mileage_mean_dct[item] = autos.loc[autos['brand'] == item, 'odometer_km'].mean()

print(brand_mileage_mean_dct)

{'volkswagen': 129041.31666341082, 'peugeot': 127270.48022598871, 'mercedes_benz': 131107.02702702704, 'ford': 124334.13317336532, 'renault': 128210.54925893636, 'audi': 129619.84069514845, 'bmw': 132845.8309482923, 'fiat': 117355.30546623794, 'opel': 129508.27559661277}


From the above, we can observe that the brands bmw, mercedes_benz, and audi have the highest mean mileage count (just like the top mean prices!).

As for the lowest mean mileage, the brands fiat, ford, and renault are the lowest (ford's mean price was a bit close to opel).

In [111]:
#converting dictionaries to Pandas Series:
brand_price_mean = pd.Series(brand_price_mean_dct)
brand_mileage_mean = pd.Series(brand_mileage_mean_dct)

#creating a dataframe from the brand_price_mean Series
brand_price_mileage_mean_df = pd.DataFrame(brand_price_mean, columns=['Mean Price'])

#assigning the brand_mileage_mean Series to the created DataFrame
brand_price_mileage_mean_df['Mean Mileage'] = brand_mileage_mean

In [121]:
brand_price_mileage_mean_df.sort_values('Mean Price', ascending=False)

Mean Price   Mean Mileage
audi           9268.353608  129619.840695
mercedes_benz  8572.600216  131107.027027
bmw            8313.271513  132845.830948
volkswagen     5478.835124  129041.316663
ford           3781.123275  124334.133173
peugeot        3095.367938  127270.480226
opel           2984.872017  129508.275597
fiat           2833.099678  117355.305466
renault        2463.199215  128210.549259

From the above (sorted by mean price) we notice that the highest prices goes for the mentioned brands before, with the highest mean mileage compared to others.

In [122]:
brand_price_mileage_mean_df.sort_values('Mean Mileage', ascending=False)

Mean Price   Mean Mileage
bmw            8313.271513  132845.830948
mercedes_benz  8572.600216  131107.027027
audi           9268.353608  129619.840695
opel           2984.872017  129508.275597
volkswagen     5478.835124  129041.316663
renault        2463.199215  128210.549259
peugeot        3095.367938  127270.480226
ford           3781.123275  124334.133173
fiat           2833.099678  117355.305466

The above table (sortd by mean mileage) also shows the same 3 brands on top.

The lowest prices however does not necessarly means a lower mileage.

Usually, a brand with a high price, and high mileage, can prove that this brand is mostly wanted. In this case, mercedes_benz is the most wanted car in the market.

# Thank You.

Hashem Ahmed.